aws-api-appsync/src/main/java/com/amplifyframework/core/model/auth/MultiAuthorizationTypeIterator.java

```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amplifyframework.core.model.auth;
+
+import androidx.annotation.NonNull;
+
+import com.amplifyframework.api.aws.AuthorizationType;
+import com.amplifyframework.core.model.AuthRule;
+import com.amplifyframework.core.model.AuthStrategy;
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+
+/**
+ * An iterator of authorization types backed by a
+ * list of auth rules. Auth rules should be sorted in which
+ * they need to be evaluated.
+ */
+public final class MultiAuthorizationTypeIterator implements AuthorizationTypeIterator {
+    private int currentIdx = 0;
+    private AuthRule effectiveRule;
+    private final List<AuthRule> authRules;
+
+    /**
+     * Constructor that takes a list of auth rules and uses the default comparator.
+     * @param authRules The list of auth rules.
+     */
+    public MultiAuthorizationTypeIterator(List<AuthRule> authRules) {
+        this(authRules, new PriorityBasedAuthRuleProviderComparator());
+    }
```

Comment:
I'll take a look, but I believe there are safeguards in place to prevent customers from setting multiple rules of the same strategy/provider combination on the same type. For instance, if more than one owner rule is added to the same type, I believe currently throw an exception. I'll double check on this though. If nothing else, I can add a check here to de-dupe where it makes sense.

Reply:
I was incorrect on that 🤦  See [here](https://docs.amplify.aws/cli/graphql-transformer/auth#multiple-authorization-rules). Allows multiple owner rules. I'll work on adding something to handle this.

Reply:
OK. I added some logic there to account for potential duplicates. One thing to keep in mind for owner and group rules:

Given the following auth rules:

```
{ allow: owner, provider:userPools }
{ allow: group, provider:userPools }
```
userPools may be tried twice: first in the context of the owner rule, which would add the owner variable to the requests; and then in the context of group in which case we wouldn't pass the owner information.